**This script contains the following points:**

    1. Importing libraries
    2. Importing data
    3. Wrangling data (Question 4)
        3.1 Renaming columns
        3.2 Changing data types  
    4. Data quality and consistency checks (Question 5)
        4.1 Missing values
        4.2 Mixed-type values
        4.3 Duplicates
    5. Combining data
        5.1 Preparing df_ords_prods for merge
            Check general structure and metadata
            Change data types and drop redundant columns
            Locate outliers and mark them NaN
        5.2 Merging dataframes (Question 6)
    6. Exporting data

# 1. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [2]:
# Turn project folder path into a string
path = r'/Users/sarahtischer/Desktop/CareerFoundry/Data Immersion/Achievement 4/01-2024_Instacart_Basket_Analysis'

In [3]:
# Import "customers.csv"
df_cust = pd.read_csv(os.path.join(path, '02_Data', 'Original_data', 'customers.csv'))

In [4]:
df_cust.shape

(206209, 10)

In [5]:
# Import "orders_products_merged(3).pkl"
df_ords_prods = pd.read_pickle(os.path.join(path, '02_Data', 'Prepared_data', 'orders_products_merged(3).pkl'))

In [6]:
df_ords_prods.shape

(32404859, 23)

# 3. Wrangling data

In [7]:
# Check structure of "customers.csv" dataframe
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [8]:
df_cust.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

## 3.1 Renaming columns

In [9]:
# Rename columns
df_cust.rename(
    columns = {'First Name' : 'first_name', 
               'Surnam' : 'last_name',
               'Gender' : 'gender',
               'STATE' : 'state',
               'Age' : 'age',
               'n_dependants' : 'dependants',
               'fam_status' : 'family_status'
              }, 
    inplace = True
)

In [10]:
# Check the output
df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants family_status  income  
0           3       married  165665  
1           0        single   59285  
2           2       married   99568  
3           0        single   42049  
4           1       married   40374

## 3.2 Changing data types

In [11]:
# Print info of df_cust
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        206209 non-null  int64 
 1   first_name     194950 non-null  object
 2   last_name      206209 non-null  object
 3   gender         206209 non-null  object
 4   state          206209 non-null  object
 5   age            206209 non-null  int64 
 6   date_joined    206209 non-null  object
 7   dependants     206209 non-null  int64 
 8   family_status  206209 non-null  object
 9   income         206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [12]:
# Change data types
df_cust['user_id'] = df_cust['user_id'].astype('str')
df_cust['age'] = df_cust['age'].astype('int8')
df_cust['dependants'] = df_cust['dependants'].astype('int8')
df_cust['income'] = df_cust['income'].astype('int32')

In [13]:
# Check the output
df_cust.dtypes

user_id          object
first_name       object
last_name        object
gender           object
state            object
age                int8
date_joined      object
dependants         int8
family_status    object
income            int32
dtype: object

# 4. Data quality and consistency checks

In [14]:
# Suppress scientific notation by specifying the display format for floating-point numbers
pd.options.display.float_format = '{:,.2f}'.format

# Print descriptive statistics
df_cust.describe()

age  dependants     income
count 206,209.00  206,209.00 206,209.00
mean       49.50        1.50  94,632.85
std        18.48        1.12  42,473.79
min        18.00        0.00  25,903.00
25%        33.00        0.00  59,874.00
50%        49.00        1.00  93,547.00
75%        66.00        3.00 124,244.00
max        81.00        3.00 593,901.00

#### *<mark>Notes</mark>*

1. Income Distribution:

* The minimum income is \$25,903, which seems reasonable.
* The maximum income is \$593,901, which is significantly higher than the 75th percentile (Q3). This could indicate potential outliers or extreme values.

2. Age Distribution:

* The minimum age is 18, which is typical for adult users.
* The maximum age is 81, and the mean age is 49.5. The maximum age doesn't seem unusual by itself, but it's essential to consider the context of the dataset and the nature of the Instacart platform.

To further investigate potential anomalies:

* Examine the distribution of income visually, considering a histogram or boxplot.
* Check for extreme values or outliers in both age and income.

## 4.1 Missing values

In [15]:
# Check for missing values
df_cust.isnull().sum()

user_id              0
first_name       11259
last_name            0
gender               0
state                0
age                  0
date_joined          0
dependants           0
family_status        0
income               0
dtype: int64

In [16]:
# Assign rows with missing values to df_nan and print the output
df_nan = df_cust[df_cust['first_name'].isnull() == True]

df_nan

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        dependants family_status  income  
53               2       married   41709  
73               0        single   82518  
82               3       married  117099  
99               1       married  155673  
105              3       married  151819  
...            ...           ...     ...  
206038           3       married   87783  
206044           2       married  108488  
206090           1       married   45275  
206162           0        single   56325  
206171           2       married   59222  

[11259 rows x 10 columns]

#### *<mark>Note:</mark> The customers dataframe contains 11259 NaN values in the column 'first_name'. However, since names are PII and shouldn't be used in the analysis, it's fine to disregard the missing values.*

## 4.2 Mixed-type values

In [17]:
# Check for mixed types
for col in df_cust.columns.tolist():
  mixed = (df_cust[[col]].map(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[mixed]) > 0:
    print (col)

first_name


In [18]:
# Unify data type of 'df_cust' to str
df_cust['first_name'] = df_cust['first_name'].astype('str')

In [19]:
# Check the output
df_cust.dtypes

user_id          object
first_name       object
last_name        object
gender           object
state            object
age                int8
date_joined      object
dependants         int8
family_status    object
income            int32
dtype: object

## 4.3 Duplicates

In [20]:
# Find full duplicates
df_dups = df_cust[df_cust.duplicated()]

df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, dependants, family_status, income]
Index: []

#### *<mark>Note:</mark> The customers dataframe doesn't contain any duplicates.*

# 5. Combining data

## 5.1 Preparing df_ords_prods for merge

#### Check general structure and metadata

In [21]:
df_ords_prods.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19    5.80   
1           1  Chocolate Sandwich Cookies        61             19    5.80   
2           1  Chocolate Sandwich Cookies        61             19    5.80   
3           1  Chocolate Sandwich Cookies        61             19    5.80   
4           1  Chocolate Sandwich Cookies        61             19    5.80   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc    busiest_days busiest_period_of_day  \
0  ...    both  Mid-range product  Regularly busy           Most orders   
1  ...    both  Mid-range product  Regularly busy        Average orders   
2  ...    both  Mid-range product     Busiest day        Average orders   
3  ...    both  Mid-range product      Least busy           Most orders   
4  ...    both  Mid-range product      Least busy        Average orders   

  max_order      loyalty_flag avg_price  spending_flag median_frequency  \
0        32  Regular customer      6.94    Low spender             8.00   
1        32  Regular customer      6.94    Low spender             8.00   
2         5      New customer      7.93    Low spender             8.00   
3         3      New customer      4.97    Low spender             9.00   
4         3      New customer      4.97    Low spender             9.00   

      frequency_flag  
0  Frequent customer  
1  Frequent customer  
2  Frequent customer  
3  Frequent customer  
4  Frequent customer  

[5 rows x 23 columns]

In [22]:
df_ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                int64   
 6   user_id                 int64   
 7   order_number            int64   
 8   orders_day_of_week      int64   
 9   order_hour_of_day       int64   
 10  days_since_prior_order  float64 
 11  add_to_cart_order       int64   
 12  reordered               int64   
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_days            object  
 16  busiest_period_of_day   object  
 17  max_order               int64   
 18  loyalty_flag            object  
 19  avg_price               float64 
 20  spending_flag           object  
 21  median

In [23]:
df_ords_prods.describe()

product_id      aisle_id  department_id        prices      order_id  \
count 32,404,859.00 32,404,859.00  32,404,859.00 32,404,859.00 32,404,859.00   
mean      25,598.66         71.20           9.92         11.98  1,710,745.50   
std       14,084.00         38.21           6.28        495.66    987,298.83   
min            1.00          1.00           1.00          1.00          2.00   
25%       13,544.00         31.00           4.00          4.20    855,947.00   
50%       25,302.00         83.00           9.00          7.40  1,711,049.00   
75%       37,947.00        107.00          16.00         11.30  2,565,499.00   
max       49,688.00        134.00          21.00     99,999.00  3,421,083.00   

            user_id  order_number  orders_day_of_week  order_hour_of_day  \
count 32,404,859.00 32,404,859.00       32,404,859.00      32,404,859.00   
mean     102,937.23         17.14                2.74              13.43   
std       59,466.10         17.54                2.09               4.25   
min            1.00          1.00                0.00               0.00   
25%       51,422.00          5.00                1.00              10.00   
50%      102,616.00         11.00                3.00              13.00   
75%      154,389.00         24.00                5.00              16.00   
max      206,209.00         99.00                6.00              23.00   

       days_since_prior_order  add_to_cart_order     reordered     max_order  \
count           30,328,763.00      32,404,859.00 32,404,859.00 32,404,859.00   
mean                    11.10               8.35          0.59         33.05   
std                      8.78               7.13          0.49         25.16   
min                      0.00               1.00          0.00          1.00   
25%                      5.00               3.00          0.00         13.00   
50%                      8.00               6.00          1.00         26.00   
75%                     15.00              11.00          1.00         47.00   
max                     30.00             145.00          1.00         99.00   

          avg_price  median_frequency  
count 32,404,859.00     32,404,854.00  
mean          11.98             10.40  
std           83.24              7.13  
min            1.00              0.00  
25%            7.39              6.00  
50%            7.82              8.00  
75%            8.25             13.00  
max       25,005.42             30.00

#### Change data types, rename columns and drop redundant columns

In [24]:
# Rename columns
df_ords_prods.rename(
    columns = {'orders_day_of_week' : 'order_day_of_week', 
               'price_range_loc' : 'price_range',
              }, 
    inplace = True
)

In [27]:
# Change data types to reduce size
df_ords_prods['product_id'] = df_ords_prods['product_id'].astype('int32')
df_ords_prods['aisle_id'] = df_ords_prods['aisle_id'].astype('int16')
df_ords_prods['department_id'] = df_ords_prods['department_id'].astype('int8')
df_ords_prods['prices'] = df_ords_prods['prices'].astype('float32')
df_ords_prods['order_id'] = df_ords_prods['order_id'].astype('int32')
df_ords_prods['user_id'] = df_ords_prods['user_id'].astype('str')
df_ords_prods['order_number'] = df_ords_prods['order_number'].astype('int8')
df_ords_prods['order_day_of_week'] = df_ords_prods['order_day_of_week'].astype('int8')
df_ords_prods['order_hour_of_day'] = df_ords_prods['order_hour_of_day'].astype('int8')
df_ords_prods['days_since_prior_order'] = df_ords_prods['days_since_prior_order'].astype('float32')
df_ords_prods['add_to_cart_order'] = df_ords_prods['add_to_cart_order'].astype('int16')
df_ords_prods['reordered'] = df_ords_prods['reordered'].astype('int8')
df_ords_prods['max_order'] = df_ords_prods['max_order'].astype('int8')
df_ords_prods['avg_price'] = df_ords_prods['avg_price'].astype('float16')
df_ords_prods['median_frequency'] = df_ords_prods['median_frequency'].astype('float16')

In [28]:
# Drop redundant variables
df_ords_prods = df_ords_prods.drop(columns = ['aisle_id', '_merge'])

In [29]:
# Check the output
df_ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 21 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int32  
 1   product_name            object 
 2   department_id           int8   
 3   prices                  float32
 4   order_id                int32  
 5   user_id                 object 
 6   order_number            int8   
 7   order_day_of_week       int8   
 8   order_hour_of_day       int8   
 9   days_since_prior_order  float32
 10  add_to_cart_order       int16  
 11  reordered               int8   
 12  price_range             object 
 13  busiest_days            object 
 14  busiest_period_of_day   object 
 15  max_order               int8   
 16  loyalty_flag            object 
 17  avg_price               float16
 18  spending_flag           object 
 19  median_frequency        float16
 20  frequency_flag          object 
dtypes: float16(2), float32(2), in

#### Locate outliers and mark them NaN

In [30]:
# Locate outliers in 'prices' column & mark them NaN
df_ords_prods.loc[df_ords_prods['prices'] > 100, 'prices'] = np.nan

In [31]:
# Check the output
df_ords_prods['prices'].max()

25.0

## 5.2 Merging dataframes

In [32]:
# Merge df_ords_prods and df_cust using 'inner' join as a full merge set is required
df_all_merged = df_ords_prods.merge(df_cust, on = ['user_id'], indicator = True)

In [33]:
# Check the output
df_all_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [34]:
df_all_merged.head()

product_id                product_name  department_id  prices  order_id  \
0           1  Chocolate Sandwich Cookies             19    5.80   3139998   
1           1  Chocolate Sandwich Cookies             19    5.80   1977647   
2         907        Premium Sliced Bacon             12   20.00   3160996   
3         907        Premium Sliced Bacon             12   20.00   2254091   
4        1000                    Apricots             10   12.90    505689   

  user_id  order_number  order_day_of_week  order_hour_of_day  \
0     138            28                  6                 11   
1     138            30                  6                 17   
2     138             1                  5                 13   
3     138            10                  5                 14   
4     138             9                  6                 12   

   days_since_prior_order  ...  first_name  last_name gender      state age  \
0                    3.00  ...     Charles        Cox   Male  Minnesota  81   
1                   20.00  ...     Charles        Cox   Male  Minnesota  81   
2                     NaN  ...     Charles        Cox   Male  Minnesota  81   
3                    6.00  ...     Charles        Cox   Male  Minnesota  81   
4                    7.00  ...     Charles        Cox   Male  Minnesota  81   

   date_joined dependants  family_status income  _merge  
0     8/1/2019          1        married  49620    both  
1     8/1/2019          1        married  49620    both  
2     8/1/2019          1        married  49620    both  
3     8/1/2019          1        married  49620    both  
4     8/1/2019          1        married  49620    both  

[5 rows x 31 columns]

In [35]:
df_all_merged.shape

(32404859, 31)

# 6. Exporting data

In [36]:
%%time
# Export df_all_merged as "orders_products_customers_merged.pkl" 
df_all_merged.to_pickle(os.path.join(path, '02_Data', 'Prepared_data', 'orders_products_customers_merged.pkl'))

CPU times: user 17.6 s, sys: 1min 13s, total: 1min 30s
Wall time: 2min 1s
